nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs

```diff
@@ -0,0 +1,843 @@
+        for zone_id in &self.in_service_clickhouse_zones.servers {
+            if !new_config.servers.contains_key(zone_id) {
+                // Allocate a new `ServerId` and map it to the server zone
+                new_config.max_used_server_id += 1.into();
+                new_config
+                    .servers
+                    .insert(*zone_id, new_config.max_used_server_id);
+            }
+        }
+
+        // Now we need to configure the keepers. We can only add or remove
+        // one keeper at a time.
+        //
+        // We need to see if we have any keeper inventory so we can compare it
+        // with our current configuration and see if any changes are required.
+        // If we fail to retrieve any inventory for keepers in the current
+        // collection than we must not modify our keeper config, as we don't
+        // know whether a configuration is ongoing or not.
+        let current_keepers: BTreeSet<_> =
+            self.parent_config.keepers.values().cloned().collect();
+        let Some(inventory_membership) = &self.inventory else {
+            return bump_gen_if_necessary(new_config);
+        };
+
+        // If we've already seen this inventory then skip keeper planning
+        if inventory_membership.leader_committed_log_index
+            <= self.parent_config.highest_seen_keeper_leader_committed_log_index
+        {
+            return bump_gen_if_necessary(new_config);
+        }
+
+        // Save the log index for next ime
+        new_config.highest_seen_keeper_leader_committed_log_index =
+            inventory_membership.leader_committed_log_index;
+
+        if current_keepers != inventory_membership.raft_config {
+            // We're still trying to reach our desired state We want to ensure
+            // however, that if we are currently trying to add a node, that we
+            // have not expunged the zone of the keeper that we are trying to
+            // add. This can happen for a number of reasons, and without this
+            // check we would not be able to make forward progress.
+            let mut added_keepers: BTreeSet<_> = current_keepers
+                .difference(&inventory_membership.raft_config)
+                .cloned()
+                .collect();
+
+            // We should only be adding or removing 1 keeper at a time
+            if added_keepers.len() > 1 {
+                error!(
+                    self.log,
+                    concat!(
+                        "Keeper membership error: attempting to add",
+                        "more than one keeper to cluster: {:?}.",
+                        "Did keeper membership change out from under us?"
+                    ),
+                    added_keepers
+                );
+                return Err(KeeperAllocationError::BadMembershipChange {
+                    added_keepers,
+                });
+            }
+
+            // If we are not adding a keeper than we are done.
```

Comment:
```suggestion
            // If we are not adding a keeper then we are done.
```

Reply:
Good catch. Fixed in [340046e](https://github.com/oxidecomputer/omicron/pull/6566/commits/340046eac678f712d813de8a6bcb7ed8fbb5435a)